File: scripts/measured_clutch_semantics.py
"""Stable measured underside-clutch identities bound to exact source seats."""

from __future__ import annotations

import re
from typing import Sequence

from measured_part_typescript_literals import numbers, string_literal

ClutchPortSemanticRow = tuple[tuple[int, int, int], str, tuple[str, ...]]
ClutchPortSemantics = tuple[ClutchPortSemanticRow, ...]

_CLUTCH_PORT_ID = re.compile(
    r"undersideClutch:[A-Za-z0-9_-]+(?::[A-Za-z0-9_-]+)*\Z"
)


def validate_clutch_port_semantics(rows: ClutchPortSemantics) -> None:
    """Refuse ambiguous identities before a measured plan can be emitted."""

    positions = [position for position, _, _ in rows]
    if len(set(positions)) != len(positions):
        raise ValueError("Measured clutch semantics must name unique exact seats.")
    port_ids = [port_id for _, port_id, _ in rows]
    if len(set(port_ids)) != len(port_ids):
        raise ValueError("Measured clutch semantics must name unique stable undersideClutch IDs.")
    for position, port_id, groups in rows:
        if not all(type(value) is int for value in position):
            raise ValueError("Measured clutch semantic seats must use whole-LDU integer positions.")
        if type(port_id) is not str or _CLUTCH_PORT_ID.fullmatch(port_id) is None:
            raise ValueError(
                "Measured clutch semantic IDs must match "
                "'undersideClutch:<non-empty colon-separated token>'."
            )
        if (
            len(set(groups)) != len(groups)
            or any(type(group) is not str or not group.strip() for group in groups)
        ):
            raise ValueError(
                "Measured clutch shared-capacity groups must be non-empty unique text per seat."
            )


def render_clutch_port_semantics(
    design_id: str,
    rows: ClutchPortSemantics,
    measured_positions: Sequence[Sequence[float]],
) -> str:
    """Align position-keyed plan rows with the measured source order and emit them."""

    by_position = {position: (port_id, groups) for position, port_id, groups in rows}
    measured = [tuple(position) for position in measured_positions]
    if len(measured) != len(rows) or set(measured) != set(by_position):
        raise ValueError(
            f"Part {design_id} measured clutch semantic seats {sorted(by_position)} "
            f"do not exactly match measured source seats {sorted(set(measured))}."
        )
    rendered: list[str] = []
    for position in measured:
        port_id, groups = by_position[position]
        fields = [
            f"positionLdu: [{numbers(position)}]",
            f"id: {string_literal(port_id)}",
        ]
        if groups:
            group_literals = ", ".join(string_literal(group) for group in groups)
            fields.append(f"sharedCapacityGroupIds: [{group_literals}]")
        rendered.append("{ " + ", ".join(fields) + " }")
    return "[" + ", ".join(rendered) + "]"

File: packages/catalog/src/measured-clutch-semantics.ts
import type {
  MeasuredClutchPortSemanticRow,
  MeasuredPartBlueprint,
} from "./measured-part-types.ts";

const CLUTCH_PORT_ID = /^undersideClutch:[A-Za-z0-9_-]+(?::[A-Za-z0-9_-]+)*$/u;

function fail(blueprint: MeasuredPartBlueprint, message: string): never {
  throw new Error(`Measured part ${blueprint.designId} (${blueprint.ldrawId}) ${message}`);
}

/** Validate generated stable clutch ids before the measured factory uses them. */
export function compileMeasuredClutchPortSemantics(
  blueprint: MeasuredPartBlueprint,
): readonly MeasuredClutchPortSemanticRow[] | undefined {
  const rows = blueprint.clutchPortSemantics;
  if (rows === undefined) return undefined;
  if (blueprint.clutchSharedCapacityGroupIds !== undefined) {
    fail(
      blueprint,
      "declares both stable clutch semantics and legacy clutch shared-capacity rows; one declaration must own each measured seat's identity and capacity claims.",
    );
  }
  if (!Array.isArray(rows) || rows.length !== blueprint.clutchesLdu.length) {
    fail(
      blueprint,
      `declares ${Array.isArray(rows) ? rows.length : "a non-array set of"} stable clutch semantic rows for ${blueprint.clutchesLdu.length} underside clutches; generated rows must stay aligned one-for-one with the exact measured source seats.`,
    );
  }

  const compiled: MeasuredClutchPortSemanticRow[] = [];
  const ids = new Set<string>();
  rows.forEach((unsafeRow, index) => {
    if (unsafeRow === null || typeof unsafeRow !== "object") {
      fail(blueprint, `stable clutch semantic row ${index} is not an object.`);
    }
    const row = unsafeRow as Partial<MeasuredClutchPortSemanticRow>;
    const position = row.positionLdu;
    const expectedPosition = blueprint.clutchesLdu[index]!;
    if (
      !Array.isArray(position) ||
      position.length !== 3 ||
      !position.every(Number.isSafeInteger) ||
      position.some((coordinate, axis) => coordinate !== expectedPosition[axis])
    ) {
      fail(
        blueprint,
        `stable clutch semantic row ${index} witnesses positionLdu ${JSON.stringify(position)} but exact measured source seat ${index} is [${expectedPosition.join(", ")}]; rows must preserve source-sorted position and identity together.`,
      );
    }
    if (typeof row.id !== "string" || !CLUTCH_PORT_ID.test(row.id)) {
      fail(
        blueprint,
        `stable clutch semantic row ${index} names id ${JSON.stringify(row.id)}; require undersideClutch:<non-empty colon-separated token>.`,
      );
    }
    if (ids.has(row.id)) {
      fail(blueprint, `stable clutch semantic row ${index} repeats connector id ${row.id}.`);
    }
    ids.add(row.id);

    const groups = row.sharedCapacityGroupIds;
    if (
      groups !== undefined &&
      (!Array.isArray(groups) ||
        groups.length === 0 ||
        new Set(groups).size !== groups.length ||
        groups.some((group) => typeof group !== "string" || group.trim().length === 0))
    ) {
      fail(
        blueprint,
        `stable clutch semantic row ${index} has sharedCapacityGroupIds ${JSON.stringify(groups)}; when present they must be non-empty unique text.`,
      );
    }
    compiled.push({
      positionLdu: [position[0], position[1], position[2]],
      id: row.id,
      ...(groups === undefined ? {} : { sharedCapacityGroupIds: groups }),
    });
  });
  return compiled;
}
